node/src/utils/ds.rs

```diff
@@ -1,18 +1,96 @@
 //! Datasize helper functions.
 
-use std::mem::size_of;
+use std::{
+    collections::{HashMap, HashSet},
+    mem::size_of,
+};
+
+use datasize::DataSize;
+use rand::{
+    rngs::StdRng,
+    seq::{IteratorRandom, SliceRandom},
+    SeedableRng,
+};
+
+/// Number of items to sample when sampling a large collection.
+const SAMPLE_SIZE: usize = 100;
 
 /// Estimate memory usage of a hashmap of keys and values each with no heap allocations.
-pub fn hash_map_fixed_size<K, V>(hashmap: &std::collections::HashMap<K, V>) -> usize {
+pub fn hash_map_fixed_size<K, V>(hashmap: &HashMap<K, V>) -> usize {
     hashmap.capacity() * (size_of::<V>() + size_of::<K>() + size_of::<usize>())
 }
 
 /// Estimate memory usage of a hashset of items with no heap allocations.
-pub fn hash_set_fixed_size<T>(hashset: &std::collections::HashSet<T>) -> usize {
+pub fn hash_set_fixed_size<T>(hashset: &HashSet<T>) -> usize {
     hashset.capacity() * (size_of::<T>() + size_of::<usize>())
```

Comment:
The `size_of::<usize>()` was probably copied from `Vec` — for hash sets and maps it's more complicated, but maybe the overhead is negligible?

Reply:
It has been copied from the estimate from `datasize` itself, unfortunately I did not look in details into the accuracy of these metrics, as that would involve very intricate knowledge of how those data structures are implemented.

Overall this is why `datasize` says "estimate" on the label.